libdrgn/program.c

```diff
@@ -667,32 +740,76 @@ struct drgn_error *drgn_program_cache_prstatus_entry(struct drgn_program *prog,
 		entry->str = data;
 		entry->len = size;
 	} else {
-		struct drgn_prstatus_map_entry entry = {
-			.value = { data, size },
-		};
-		struct drgn_error *err = get_prstatus_pid(prog, data, size,
-							  &entry.key);
+		struct drgn_thread thread = { .prog = prog, .prstatus = { data, size } };
+		struct drgn_error *err = get_prstatus_pid(prog, data, size, &thread.tid);
 		if (err)
 			return err;
-		if (drgn_prstatus_map_insert(&prog->prstatus_map, &entry,
-					     NULL) == -1)
+		*ret = thread.tid;
+		if (drgn_thread_set_insert(&prog->thread_set, &thread, NULL) == -1)
 			return &drgn_enomem;
 	}
 	return NULL;
 }
 
-static struct drgn_error *drgn_program_cache_prstatus(struct drgn_program *prog)
+static struct drgn_error *
+drgn_program_kernel_coredump_cache_crashed_thread(struct drgn_program *prog)
+{
+	assert((prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL) &&
+	       !(prog->flags & DRGN_PROGRAM_IS_LIVE));
+	assert(prog->threads_cached);
+	if (prog->crashed_thread)
+		return NULL;
+	struct drgn_error *err;
+	uint32_t crashed_thread_tid;
+	int64_t crashing_cpu_value;
+	struct drgn_object crashing_cpu;
+	drgn_object_init(&crashing_cpu, prog);
+	err = drgn_program_find_object(prog, "crashing_cpu", NULL,
+				       DRGN_FIND_OBJECT_VARIABLE,
+				       &crashing_cpu);
+	if (err)
+		goto out;
+	err = drgn_object_read_signed(&crashing_cpu, &crashing_cpu_value);
+	if (err)
+		goto out;
+	if (crashing_cpu_value < 0 ||
+	    crashing_cpu_value >= prog->prstatus_vector.size) {
+		err = drgn_error_create(
+			DRGN_ERROR_LOOKUP,
+			"crashed thread not found, debug info is possibly corrupted");
+		goto out;
+	}
+	struct nstring *prstatus =
+		&prog->prstatus_vector.data[crashing_cpu_value];
+	err = get_prstatus_pid(prog, prstatus->str, prstatus->len,
+			       &crashed_thread_tid);
+	if (err)
+		goto out;
+	err = drgn_program_find_thread(prog, crashed_thread_tid,
+				       &prog->crashed_thread);
```

Comment:
I haven't looked at the rest, but I noticed one more corner case: this won't work for PID 0. PID 0 is the idle thread that the CPU switches to if there's nothing else to run. We've seen it crash in production, usually because of bad hardware.

`find_task(0)` doesn't work because (1) PID 0 is not included in the radix tree/pid_hash and (2) there is actually a separate `task_struct` with pid == 0 per CPU. adfb04579b618b6838d0cfc6ea81293f7aee10a0 adds a helper you can use to get the idle thread by the CPU number instead. I'll merge that once tests pass.

Reply:
One sec, I'm going to rename that helper to `idle_task()` (from `idle_thread()`).

Reply:
I've now taken this into account. Do you happen to have a kernel dump where this is the case that I could test on?

Reply:
> One sec, I'm going to rename that helper to `idle_task()` (from `idle_thread()`).

No worries, will just re-rebase real quick once you've pushed.